### PR TITLE
OpenACC Parallel Constructs for MDRange Policy

### PR DESCRIPTION
**Pull request by seyonglee**
Add OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp    
Add OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp
Add KOKKOS_IF_ON_DEVICE and KOKKOS_IF_ON_HOST definitions for the OpenACC backend (Kokkos_Macros.hpp).
Implement device_abort() function for the OpenACC backend (impl/Kokkos_Error.hpp).

### CODE REVIEW

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp
Comment (Rombur) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp:61]: Can you improve the error message.

**Review by Rombur — APPROVED** [review]

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (JBludau) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:241]: Please say that only 1<ranks<7 are supported. Rank=1 should trigger the static assert too.

File: core/src/Kokkos_Macros.hpp
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:590]: This does not actually compile

Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:591]: This code is for other OpenACC compilers. In OpenACC, the only portable way to check whether a code is running on a host or a device is to use `acc_on_device` function, which can be converted into a compile-time constant by the OpenACC compiler. Unfortunately, this compile-time conversion is not guaranteed even if the input argument is a compile-time constant, and thus this code may not work in some OpenACC compilers (just best-effort approach for now).

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:591]: Do we know of any compiler that supports it?

File: core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelReduce_MDRange.hpp]: Updated the error message.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp]: Updated.

File: core/src/Kokkos_Macros.hpp
Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:591]: I have not checked it; the OpenACC spec says that it can be evaluated to a constant at compile time, but it is optional.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (dalg24) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:118]: No, I really meant that comment.  If this else-clause gets compiled there is a bug, it is not a matter of being supported or not.

File: core/src/Kokkos_Macros.hpp
Comment (dalg24) [inline core/src/Kokkos_Macros.hpp:576]: I suspect that this means the fallback to host if no device is detected won't actually work

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (masterleinad) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:118]: Did you mean
```suggestion
      static_assert(false,
          "Kokkos Error: not supported memory layout for the OpenACC backend");
```
? Any string implicitly converts to `true`.

Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp]: Thank you for catching this; updated.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:119]: Updated the comment.

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp]: Changed `static_assert(false, …)` to `Kokkos::abort(…)` since `static_assert(false, …)` will always complain errors.

File: core/src/Kokkos_Macros.hpp
Comment (seyonglee) [inline core/src/Kokkos_Macros.hpp:497]: Setting the default device type to the OpenACC device is temporarily disabled, since there are many tests running on the default device, which will fail if using features not implemented in the OpenACC backend.
This default device type setting for the OpenACC backend will be re-enabled after all the basic features are implemented.

File: core/src/Kokkos_Macros.hpp
Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:577]: There seems to be a bug in `__builtin_is_device_code`, which is not allowed to be called on host (https://forums.developer.nvidia.com/t/nvc-up-to-22-7-undefined-reference-to-builtin-is-device-code-for-empty-if-target-paths/223232). Therefore, we will have to either wait for the next release or use `acc_on_device` API.

File: core/src/Kokkos_Macros.hpp
Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:497]: We absolutely need it for View accesses.  The incremental only pass because `Kokkos::abort` does nothing on the device but the standard output is cluttered with memory access violation error messages.

Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:497]: I created another local branch, which enables the default OpenACC execution space and disable tests evaluating not-implemented OpenACC features. the new branch executes the incremental tests correctly. I will submit the new PR after the this PR (#5335) is merged.

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:497]: I would do it the other way around.  They are many changes on this PR that are controversial.

Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:497]: What you mean is to submit a new PR fixing on the default OpenACC execution space before this PR (#5335)?

Comment (dalg24) [reply core/src/Kokkos_Macros.hpp:497]: Yes

Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp:497]: Submitted a new PR (#5360) for the default OpenACC execution space.

**Review by pedrovalerolara — APPROVED** [review]

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (seyonglee) [inline core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:120]: OpenACC says that the tile size should be a constant positive integer expression, but `tile0` and `tile1` are not constant.

**Comment by Rombur** [comment]
The PR needs to be rebased on develop to fix the conflict

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (dalg24) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:120]: runtime or compile-time constant?

**Comment by seyonglee** [comment]
> The PR needs to be rebased on develop to fix the conflict

Rebased; thanks!

File: core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp
Comment (seyonglee) [reply core/src/OpenACC/Kokkos_OpenACC_ParallelFor_MDRange.hpp:120]: compile-time constant.

File: core/src/Kokkos_Macros.hpp
Comment (seyonglee) [reply core/src/Kokkos_Macros.hpp]: PR #5360 is merged into the develop branch and also merged into this PR as a part of rebasing.

**Review by JBludau — APPROVED** [review]
looks good

**Review by dalg24 — CHANGES_REQUESTED** [review]
Superseded by #5376

**Pull request closed without merging**
